### PR TITLE
feat(thanos): add support for named stores

### PR DESCRIPTION
**Pull request by ashwanthgoli**
**What this PR does / why we need it**:

This pr adds named store support for thanos based object clients. `-use-thanos-objstore` flag has to be set to true to use these configs.

Below is an example of how to configure a named s3 store. `store-1` can then be set as the `object_store` in your [period_configs](https://grafana.com/docs/loki/latest/configure/#period_config)
```
storage_config:
  object_store:
    named_stores:
      s3:
        store-1:
          endpoint: s3://foo-bucket
          region: us-west1

```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/bucket/object_client_adapter.go [hunk 1]
```diff
@@ -19,9 +24,36 @@ type ObjectClientAdapter struct {
...
+		storeCfg  = cfg.Config
+	)
+
+	if st, ok := cfg.NamedStores.LookupStoreType(backend); ok {
+		storeType = st
+		// override config with values from named store config
+		if err := cfg.NamedStores.OverrideConfig(&storeCfg, backend, storeType); err != nil {
+			return nil, err
+		}
+	}
```
Comment (JoaoBraveCoding) [inline pkg/storage/bucket/object_client_adapter.go:39]: I'm a bit confused by this part. We set `storeType` using `backend` but then also use `backend` to search for a namedStore.

File: pkg/storage/bucket/client.go [hunk 2]
```diff
@@ -47,87 +48,79 @@ var (
...
+// Config holds configuration for accessing long-term storage.
+type Config struct {
 	// Backends
 	S3         s3.Config         `yaml:"s3"`
 	GCS        gcs.Config        `yaml:"gcs"`
 	Azure      azure.Config      `yaml:"azure"`
 	Swift      swift.Config      `yaml:"swift"`
 	Filesystem filesystem.Config `yaml:"filesystem"`
 
+	StoragePrefix string `yaml:"storage_prefix"`
```
[the remark this answers is not in the document]
Comment (ashwanthgoli) [reply pkg/storage/bucket/client.go:61]: hmm, we still use [it](https://github.com/grafana/loki/blob/d4cc37697da028e2ec9f13aa39eca76504d0f4c1/pkg/storage/bucket/client.go#L189) when prefix is configured

File: pkg/storage/bucket/named_stores.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	storeType, ok := ns.LookupStoreType(namedStore)
+	if !ok {
+		return fmt.Errorf("Unrecognized named storage config %s", namedStore)
+	}
+
+	switch storeType {
+	case GCS:
+		nsCfg, ok := ns.GCS[namedStore]
+		if !ok {
+			return fmt.Errorf("Unrecognized named s3 storage config %s", namedStore)
```
Comment (JoaoBraveCoding) [inline pkg/storage/bucket/named_stores.go:112]:
```suggestion
			return fmt.Errorf("Unrecognized named gcs storage config %s", namedStore)
```

File: pkg/storage/bucket/named_stores.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+		}
+
+		storeCfg.GCS = (gcs.Config)(nsCfg)
+	case S3:
+		nsCfg, ok := ns.S3[namedStore]
+		if !ok {
+			return fmt.Errorf("Unrecognized named gcs storage config %s", namedStore)
```
Comment (JoaoBraveCoding) [inline pkg/storage/bucket/named_stores.go:119]:
```suggestion
			return fmt.Errorf("Unrecognized named s3 storage config %s", namedStore)
```

File: pkg/storage/factory.go [hunk 5]
```diff
@@ -489,6 +515,7 @@ func NewChunkClient(name, component string, cfg Config, schemaCfg config.SchemaC
 			return client.NewClientWithMaxParallel(c, nil, 1, schemaCfg), nil
 		}
 
+	case cfg.UseThanosObjstore:
```
Comment (JoaoBraveCoding) [inline pkg/storage/factory.go:518]: Seems to be unused
```suggestion

```

**Review by salvacorts — APPROVED** [review]

**Pull request merged**